### PR TITLE
update feature: support printf with %s,%p, %d

### PR DESCRIPTION
**Pull request by vuhailongkl97**
I have just updated a printf function, I think it is useful.
Please consider to merge it.

### CODE REVIEW

**Comment by nguyenzung** [comment]
Thanks for your update!

It is very helpful,  i see that you use std lib. Because std lib invoke unexpected function in standard library, and some of them relate to the dynamic memory that i implemented. I am not sure it is suitable and don't make memory leak. So for now, i just use pure C++ without any lib.

File: include/kernel/printer.h [hunk 1]
```diff
@@ -23,7 +39,59 @@ class Printer
...
+    template <typename T, typename... Args>
+    static void printfHelper(int i, const char *format, T first, Args... args);
 };
 
+template <typename T, typename... Args>
+void Printer::printfHelper(int i, const char *format, T first, Args... args) {
+    char tmp[10] = {0};
+
+    for (; format[i] != '\0'; i++) {
+        if (format[i] == '%' && format[i + 1] != '\0') {
```
Comment (nguyenzung) [inline include/kernel/printer.h:59]: Sometime, we init a string by using new operator. In this case, the terminated character is not '\0'

**Comment by vuhailongkl97** [comment]
I use the STL header for the template, it's not related to any libraries in runtime due to template(deduction) is a feature of the compiler in compile time. in my case, STL only uses pure c++ features.

**Comment by nguyenzung** [comment]
> I use the STL header for the template, it's not related to any libraries in runtime due to template(deduction) is a feature of the compiler in compile time. in my case, STL only uses pure c++ features.

Thanks! Your feature is really useful.

**Pull request merged**